pyo3-derive-backend/src/py_class.rs

```diff
@@ -133,9 +152,44 @@ fn impl_class(
         quote! {0}
     };
 
+    // If there are no type arguments, simply use a static variable.
+    let type_object_impl = if generics.params.is_empty() {
+        quote! {
+            static mut TYPE_OBJECT: ::pyo3::ffi::PyTypeObject = ::pyo3::ffi::PyTypeObject_INIT;
+            &mut TYPE_OBJECT
+        }
+    }
+    // If there are t-args, lacking support for t-arg dependent statics in Rust,
+    // we fall back to using a global hash map for resolving the correct type
+    // object using TypeId.
+    else {
+        quote! {
+            use std::collections::hash_map::Entry;
+            let mut map = ::pyo3::typeob::PY_TYPE_OBJ_MAP.lock().unwrap();
```

Comment:
As @konstin commented type-object map should be declared here.
I don't think it's good to use just one hash map for every classes.

Reply:
I initially had it with one map per class, however it comes with the ugly drawback that users will have to directly depend on `lazy_static` and `#[macro_use] extern crate` it in their own code as soon as they use generic structs. Also, the error message when forgetting it is really unhelpful. Maybe I can re-export lazy_static as part of PyO3's interface to counter this. Will check.

Reply:
So how about just using `static mut TYPE_OBJ_MAP = Mutex<HashMap<..>>`?

Reply:
Static initializers are required to be constant, which neither Mutex' nor HashMap's constructors are, unfortunately. See [this playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=131285354e191551db5645439dc4f693). Lazy static defers initializer execution to RT, thus allowing this to work.

Reply:
Ah, sorry, I understand.
I think
```rust
static mut TYPE_OBJ_MAP: *mut Mutex<HashMap<..>> = ..;
static INIT_TYPE_MAP: sync::Once = sync::ONCE_INIT;
INIT_TYPE_MAP.call_once(||{
...
});
```
works, but please use your favorite way.